components/UserList.js

```diff
@@ -1,37 +1,42 @@
 import { useRouter } from 'next/router'
+import { stringify } from 'qs'
 import { useState } from 'react'
 import { Button, ListGroup } from 'react-bootstrap'
-import useSWR, { mutate } from 'swr'
+import useSWR from 'swr'
 import { useAuth } from 'use-auth0-hooks'
 
 import PageControls from './PageControls'
 import UserRow from './UserRow'
 import { AUTH0_SCOPE, USER_TYPES } from '../util/constants'
-import { secureFetch, secureFetchHandleErrors } from '../util/middleware'
-
-function _getUrl (type) {
-  const selectedType = USER_TYPES.find(t => t.value === type)
-  if (!selectedType) throw new Error(`Type: ${type} does not exist!`)
-  return selectedType.url
-}
+import { getUserUrl, secureFetch } from '../util/middleware'
 
 /**
  * This component renders a list of users (can be any subtype of otp-middleware's
  * AbstractUser).
  */
-function UserList ({ summaryView, type }) {
+function UserList ({ fetchUsers, summaryView, type, updateUser }) {
+  // Set up hooks, state.
   const { accessToken, isAuthenticated } = useAuth({
     audience: process.env.AUTH0_AUDIENCE,
     scope: AUTH0_SCOPE
   })
   const [offset, setOffset] = useState(0)
   const router = useRouter()
+  // Ensure user is authenticated and type from query param is valid.
+  const selectedType = USER_TYPES.find(t => t.value === type)
+  if (!isAuthenticated) return null
+  if (!selectedType) return <div>Page does not exist!</div>
+  // Fetch user data.
+  const limit = 10
+  const url = `${getUserUrl(type)}?${stringify({limit, offset})}`
+  const getAllResult = useSWR(url)
+  const { data, error, mutate: mutateList } = getAllResult
+  const users = data && data.data
+  // Set up on click handlers with mutates to trigger refresh on updates.
```

Comment:
The contents of these click handlers are beginning to like too much application logic being put in a presentation layer. Furthermore, these mutations, state updates etc might need to be reused elsewhere. I'm not super familiar with useSWR and the state updates, but it seems like all these could be made more reusable and stored in some kind of file that manages state, fetching, etc. Something like this: https://swr.vercel.app/getting-started#make-it-reusable

Reply:
See #32.